### PR TITLE
Refactor internal http client usage

### PR DESCRIPTION
**Pull request by wendigo**
The previous approach was dangerous, as it was modifying configuration for all HTTP clients and adding global filters even for HTTP clients there were not supposed to be used for internal communication.

With this change, it's now intentional which HTTP clients are supposed to use secured internal communication.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by wendigo** [comment]
@electrum PTAL. I think that I got this change right

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        });
+
+        // Passes internal token to http client
+        httpClientBindingBuilder.addFilterBinding().to(InternalAuthenticationManager.class);
+
+        if (withTracing) {
+            httpClientBindingBuilder.withTracing();
+        }
+
+        if (!filters.isEmpty()) {
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java:72]: nit: redundant if, just `filters.forEach`

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            filters.forEach(httpClientBindingBuilder::withFilter);
+        }
+    }
+
+    static void configureClient(HttpClientConfig httpConfig, InternalCommunicationConfig internalCommunicationConfig)
```
Comment (lukasz-walkiewicz) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java:77]: Too broad access? x3

**Review by lukasz-walkiewicz — APPROVED** [review]
LGTM

File: core/trino-main/src/main/java/io/trino/server/CoordinatorModule.java [hunk 3]
```diff
@@ -206,12 +206,12 @@ protected void setup(Binder binder)
...
         binder.bind(ClusterMemoryManager.class).in(Scopes.SINGLETON);
         binder.bind(ClusterMemoryPoolManager.class).to(ClusterMemoryManager.class).in(Scopes.SINGLETON);
-        httpClientBinder(binder).bindHttpClient("memoryManager", ForMemoryManager.class)
+        install(internalHttpClientModule("memoryManager", ForMemoryManager.class)
                 .withTracing()
                 .withConfigDefaults(config -> {
                     config.setIdleTimeout(new Duration(30, SECONDS));
                     config.setRequestTimeout(new Duration(10, SECONDS));
-                });
+                }).build());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/CoordinatorModule.java:214]: nit: formatting:
```
  })
  .build();
```

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        this.annotation = requireNonNull(annotation, "annotation is null");
+        this.withTracing = withTracing;
+        this.configDefaults = requireNonNull(configDefaults, "configDefaults is null");
+        this.filters = ImmutableList.copyOf(requireNonNull(filters, "filters is null"));
+    }
+
+    @Override
+    protected void setup(Binder binder)
+    {
+        // Set default config for this given HttpClient
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java:57]: nit: I would remove this comment

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        HttpClientBinder.HttpClientBindingBuilder httpClientBindingBuilder = httpClientBinder(binder).bindHttpClient(clientName, annotation);
+        InternalCommunicationConfig internalCommunicationConfig = buildConfigObject(InternalCommunicationConfig.class);
+        httpClientBindingBuilder.withConfigDefaults(httpConfig -> {
+            configureClient(httpConfig, internalCommunicationConfig);
+            configDefaults.accept(httpConfig);
+        });
+
+        // Passes internal token to http client
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java:65]: nit: I would remove this comment

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        this.withTracing = withTracing;
+        this.configDefaults = requireNonNull(configDefaults, "configDefaults is null");
+        this.filters = ImmutableList.copyOf(requireNonNull(filters, "filters is null"));
+    }
+
+    @Override
+    protected void setup(Binder binder)
+    {
+        // Set default config for this given HttpClient
+        HttpClientBinder.HttpClientBindingBuilder httpClientBindingBuilder = httpClientBinder(binder).bindHttpClient(clientName, annotation);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java:58]: nit: static import

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    private static void configureClientForAutomaticHttps(HttpClientConfig httpConfig, InternalCommunicationConfig internalCommunicationConfig)
+    {
+        String sharedSecret = internalCommunicationConfig.getSharedSecret()
+                .orElseThrow(() -> new IllegalArgumentException("Internal shared secret must be set when internal HTTPS is enabled"));
+
+        httpConfig.setHttp2Enabled(internalCommunicationConfig.isHttp2Enabled());
+        httpConfig.setAutomaticHttpsSharedSecret(sharedSecret);
+    }
+
+    private static void configureClientForManualHttps(HttpClientConfig httpConfig, InternalCommunicationConfig internalCommunicationConfig)
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java:94]: Do we still support it? I think we should deprecate it and remove eventually.

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationModule.java [hunk 8]
```diff
@@ -40,59 +39,48 @@
...
+            discoveryFilterBinder.addBinding().to(DiscoveryEncodeAddressAsHostname.class);
         }
-
+        discoveryFilterBinder.addBinding().to(InternalAuthenticationManager.class);
         binder.bind(InternalAuthenticationManager.class);
-        httpClientBinder(binder).bindGlobalFilter(InternalAuthenticationManager.class);
     }
 
-    private static URI toIpEncodedAsHostnameUri(URI uri)
+    private static class DiscoveryEncodeAddressAsHostname
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationModule.java:55]: I would extract this filter as preparatory PR

**Review by kokosing — COMMENTED** [review]
LGTM % comments

File: core/trino-main/src/main/java/io/trino/server/CoordinatorModule.java (hunk 3 above)
Comment (electrum) [reply core/trino-main/src/main/java/io/trino/server/CoordinatorModule.java:214]: I think these are fine

File: core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    {
+        httpConfig.setHttp2Enabled(internalCommunicationConfig.isHttp2Enabled());
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/InternalCommunicationHttpClientModule.java:96]: As mentioned above

**Review by electrum — APPROVED** [review]

**Review by kokosing — APPROVED** [review]

**Pull request merged**